chore(*): use Yarn for dependency management

petebacondarwin opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**



**What is the current behavior? (You can also link to an open issue here)**



**What is the new behavior (if this is a feature change)?**



**Does this PR introduce a breaking change?**



**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

petebacondarwin committed:
style(*): fix `no-useless-escape` eslint errors

Narretz commented:
Node 6, here we come!
Can we have an explicite mention of the upgrade to node 6 in one of the commit messages? Is it implied by the change to yarn?

gkalpak commented on scripts/validate-commit-msg.js:
```diff
@@ -16,7 +16,7 @@ var util = require('util');
 
```
Is this file used any more (with `init-repo.sh` gone)? We could as well get rid of it altogether.

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -22,10 +22,10 @@ var ISO_DATE_REGEXP = /^\d{4,}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+(?:[+-
 //   8. Query
 //   9. Fragment
 //                 1111111111111111 222   333333    44444        555555555555555555555555    666     77777777     8888888     999
```
Nit: The last `5` should be a space.

gkalpak commented on scripts/jenkins/set-node-version.sh:
```diff
@@ -3,5 +3,7 @@
+# Use version of node.js found in .nvmrc
+nvm install
+
+npm install -g yarn
```
Nit: This file now does more than its name suggests. Maybe rename it to something like `setup-node-env`?

gkalpak reviewed (commented):
:tada: BTW, we also need to update `docs/content/misc/contribute.ngdoc`.

petebacondarwin commented:
I'll add better commit messages. I am pretty sure that yarn requires node 6+.

petebacondarwin committed:
docs(*): switch from npm to yarn

petebacondarwin committed:
chore(npm-bundle-deps): remove unused script

This became reduntant as of 507ee2d9bac77841681282dfcba9297dda576b41

petebacondarwin committed:
chore(bower): set clone depth to speed up releases

We already use `git clone ... --depth=1` for the `code.angularjs.org`
repository, which speeds up the cloning of the repository.

petebacondarwin committed:
chore(saucelabs): update sauce-connect to 4.4.1

mgol commented on Gruntfile.js:
```diff
@@ -322,13 +318,6 @@ module.exports = function(grunt) {
   });
 
-  // global beforeEach task
-  if (!process.env.TRAVIS) {
```
Shouldn't we run `yarn` here?

petebacondarwin replied:
OK, that sounds right.

mgol commented on changelog.js:
```diff
@@ -1,208 +0,0 @@
-#!/usr/bin/env node
-
```
Does this removal have anything to do with switching to Yarn? Looks like a general cleanup.

petebacondarwin replied:
Yes, you're right, I will move it into separate commits

mgol commented on src/ngComponentRouter/Router.js:
```diff
@@ -10,11 +10,11 @@
  * @installation
  *
- * Currently, the **Component Router** module must be installed via `npm`, it is not yet available
+ * Currently, the **Component Router** module must be installed via `npm`/`yarn`, it is not yet available
```
Unrelated, but this "yet" is a little reaching... :) I'd remove the word.

petebacondarwin replied:
I'll do so in a separate commit :-)

mgol commented on package.json:
```diff
@@ -9,8 +9,8 @@
   "engines": {
-    "node": "<5",
-    "npm": "~2.5"
+    "node": "^6.9.0",
```
Maybe `^6.9.1`? 6.9.1 was released a day after 6.9.0 and contained critical fixes.

petebacondarwin replied:
OK

petebacondarwin committed:
docs(ComponentRouter): it is unlikely ever to appear on bower or CDN

gkalpak commented on Gruntfile.js:
```diff
@@ -322,13 +318,6 @@ module.exports = function(grunt) {
 
-  // global beforeEach task
-  if (!process.env.TRAVIS) {
-    grunt.task.run('shell:npm-install');
```
FWIW, I think this was kind of useful...
<sub>(I can leave without it though.)</sub>

gkalpak reviewed (commented):
Added a few more nits :smiley:

petebacondarwin committed:
chore(jenkins): remove path to grunt

grunt is installed globally on jenkins so we can just use it directly.

gkalpak commented on scripts/jenkins/release.sh:
```diff
@@ -38,8 +38,7 @@ function build {
 
-  yarn
-  node_modules/.bin/grunt  ci-checks package --no-color
+  grunt  ci-checks package --no-color
```
Two spaces between 'grunt' and 'ci' 😱

gkalpak commented on scripts/jenkins/build.sh:
```diff
@@ -21,22 +21,21 @@ rm -f angular.js.size
 # BUILD #
-yarn
-node_modules/.bin/grunt ci-checks package --no-color
+grunt ci-checks package --no-color
```
Is grunt available globally?

gkalpak commented on Gruntfile.js:
```diff
@@ -292,10 +292,9 @@ module.exports = function(grunt) {
-      'npm-install': {
-        command: 'node scripts/npm/check-node-modules.js'
+      'install-node-dependencies': {
+        command: 'yarn'
```
Won't this take more time that simply comparing two files (s was happening before). I am not very keen on sacrificing speed for _every_ task for this 😕

mgol replied:
It shouldn't be a problem. The following was run on the angular.js repo with npm & yarn cache cleared:
```
$ time yarn --ignore-scripts
yarn --ignore-scripts  28.22s user 21.07s system 125% cpu 39.178 total
$ time yarn --ignore-scripts
yarn --ignore-scripts  6.49s user 2.68s system 120% cpu 7.583 total
$ time yarn --ignore-scripts
yarn --ignore-scripts  0.87s user 0.07s system 70% cpu 1.321 total
$ time yarn --ignore-scripts
yarn --ignore-scripts  0.86s user 0.07s system 70% cpu 1.313 total
```

gkalpak replied:
Right :+1:

When trying this myself, I stopped after the 2nd time :smiley:

mgol replied:
Could you try? It'd be good to make sure it's that quick not only on *NIX-es.

gkalpak replied:
I did try of course. I didn't take your word for it 😛
It "stabilizes" at 2secs on my machine, which is OK (for me at least).

mgol replied:
I wonder why it's not that fast on the second try, though. I'd expect it to behave identically in all situations where cache is populated and dependencies are correct so something looks wrong here.

mgol replied:
Let's see: https://github.com/yarnpkg/yarn/issues/2054

petebacondarwin committed:
chore(utils): install npm-run to simplify scripts

gkalpak commented on scripts/travis/build.sh:
```diff
@@ -6,19 +6,19 @@ export BROWSER_STACK_ACCESS_KEY=`echo $BROWSER_STACK_ACCESS_KEY | rev`
-node_modules/.bin/grunt tests:docs --browsers="$BROWSERS" --reporters=dots
+grunt test:promises-aplus
+grunt test:unit --browsers="$BROWSERS" --reporters=dots
+grunt tests:docs --browsers="$BROWSERS" --reporters=dots
```
Bad indentation 😛

mgol commented on i18n/generate.sh:
```diff
@@ -3,14 +3,13 @@
 set -e
 
 BASE_DIR=`dirname $0`
-cd $BASE_DIR
```
This reminded me that scripts like this one are dangerous unless you add `set -u` at the beginning. Otherwise if suddenly a variable like that gets unset and later used, you may get unexpected results.

mgol commented on scripts/jenkins/init-node.sh:
```diff
@@ -0,0 +1,9 @@
+# Use version of node.js found in .nvmrc
+nvm install
+
+npm install -g yarn
```
We should install a version matching what we have in package.json's `engines.yarn` field.

mgol replied:
[jq](https://stedolan.github.io/jq/) is a good tool for that. You can play with it at https://jqplay.org/

mgol replied:
Some environments, like Heroku, use `engines` to decide what version to use. I guess it won't happen in our situation but we should still specify the version somewhere and it doesn't hurt to do it in `engines` and read from there; if someone wanted to do sth on one of these engines-reading tools, they could use these settings.

As for humans, we could always add a check for an installed version and require it to be not older than whatever's in `engines.yarn`.

Daniel15 replied:
Please avoid installing Yarn via npm if possible. This is significantly slower than all other methods of installing Yarn, and often results in weird issues around Yarn's dependencies. We (Yarn team) do not recommend it

If running on Debian or Ubuntu, you should use Yarn's Debian repository. Otherwise, use the installation script (`curl -o- -L https://yarnpkg.com/install.sh | bash`) or just extract the tarball manually. Both methods (Debian package repository and installation script) allow you to specify the exact version of Yarn to install.

petebacondarwin committed:
chore(validate-commit): remove redundant scripts

petebacondarwin committed:
chore(grunt): run gulp directly rather than through npm

petebacondarwin commented on docs/content/misc/contribute.ngdoc:
```diff
@@ -62,7 +69,7 @@ cd angular.js
 
 # Install node.js dependencies:
-npm install
+yarn
```
Hmm, the Yarn docs generally recommend just `yarn` but OK.

petebacondarwin commented on docs/content/misc/contribute.ngdoc:
```diff
@@ -62,7 +69,7 @@ cd angular.js
+yarn
 
 # Install bower components:
 bower install
```
removed.

petebacondarwin committed:
docs(FAQ) add link to security guide

gkalpak commented on .nvmrc:
```diff
@@ -1,2 +1 @@
-4
-
+6.9.1
```
Isn't that too restrictive?

petebacondarwin replied:
6.9 is the LTS version of node. There are actually only two versions node of 6.9: 6.9.0 and 6.9.1 The former has issues that need to be fixed in 6.9.1. I don't see any reason why we should let people use versions below 6.9.

petebacondarwin replied:
Moreover `nvm` is not capable of parsing `^6.9.1` so we would be stuck with `6.9` or worse `6`, which could allow versions that we don't want such as 6.8.x and 6.9.0.

gkalpak replied:
But it means that we always need to have (and switch to) 6.9.1 locally, even when 6.9.2+ comes out 😞

gkalpak commented on Gruntfile.js:
```diff
@@ -322,13 +333,10 @@ module.exports = function(grunt) {
     }
   });
 
-  // global beforeEach task
```
You could leave this.

petebacondarwin replied:
OK, although I found it slightly confusing due to its similarity to testing terminology - like it was something special. How about I modify it to something more humane?

gkalpak commented on docs/content/misc/contribute.ngdoc:
```diff
@@ -24,20 +24,27 @@ Before you can build AngularJS, you must install and configure the following dep
+  development web server, run tests, and generate distributable files. Depending on your system,
+  you can install Node either from source or as a pre-packaged bundle.
+
+  We recommend using [nvm](https://github.com/creationix/nvm) to manage and install Node.js,
```
I've been using https://github.com/coreybutler/nvm-windows for some time now and it works fine. I think it is worth mentioning it too.

gkalpak commented on docs/content/misc/contribute.ngdoc:
```diff
@@ -24,20 +24,27 @@ Before you can build AngularJS, you must install and configure the following dep
+  We recommend using [nvm](https://github.com/creationix/nvm) to manage and install Node.js,
+  which makes it easy to change the version of Node.js per project.
+
+* [Yarn](https://yarnpkg.com): We use Yarn to install our Node.js module dependencies (rather than using npm).
```
It might be worth mentioning the installation command (`npm install -g yarn`).

Daniel15 replied:
@gkalpak - It's better to link to https://yarnpkg.com/en/docs/install. Installing Yarn via npm is not the recommended installation method and should only be used in cases where a proper package is not available for your environment.

gkalpak commented on package.json:
```diff
@@ -9,16 +9,13 @@
   "engines": {
-    "node": "<5",
-    "npm": "~2.5"
+    "node": "6.9.1",
```
Why restrict this to `6.9.1`?

gkalpak commented on scripts/travis/build.sh:
```diff
@@ -14,9 +14,9 @@ elif [ "$JOB" == "unit" ]; then
-  grunt tests:docs --browsers="$BROWSERS" --reporters=dots
+grunt test:promises-aplus
+grunt test:unit --browsers="$BROWSERS" --reporters=dots
+grunt tests:docs --browsers="$BROWSERS" --reporters=dots
```
Bad indentation (this one never gets old 😛 )

petebacondarwin replied:
Just keeping you interested :-P

gkalpak reviewed (approved):
A couple of minor comments/questions.
LGTM

gkalpak commented:
BTW, I have tested this on Windows (10) and Ubuntu (16.04) with `yarn && grunt test` and everything works fine.

Daniel15 commented on src/ng/directive/input.js:
```diff
@@ -21,11 +21,11 @@ var ISO_DATE_REGEXP = /^\d{4,}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+(?:[+-
 //   9. Fragment
-//                 1111111111111111 222   333333    44444        555555555555555555555555    666     77777777     8888888     999
-var URL_REGEXP = /^[a-z][a-z\d.+-]*:\/*(?:[^:@]+(?::[^@]+)?@)?(?:[^\s:/?#]+|\[[a-f\d:]+\])(?::\d+)?(?:\/[^?#]*)?(?:\?[^#]*)?(?:#.*)?$/i;
+//                 1111111111111111 222   333333    44444        55555555555555555555555     666     77777777     8888888     999
```
The lack of named capture groups in JavaScript regular expressions is really annoying :P

Daniel15 commented:
Please avoid installing Yarn via npm if possible. This is not recommended. It's significantly slower than all other methods of installing Yarn, and often results in weird issues around Yarn's dependencies.

If running on Debian or Ubuntu, you should use Yarn's Debian repository. Otherwise, use the installation script (`curl -o- -L https://yarnpkg.com/install.sh | bash`) or just extract the tarball manually. Both methods (Debian package repository and installation script) allow you to specify the exact version of Yarn to install.

petebacondarwin commented:
@Daniel15 - thanks for the feedback about installing Yarn. Do you have any input on how we could lock down the version of Yarn that we are using to install our dependencies?

Daniel15 commented:
It depends on how you're installing Yarn. If you use the Debian repo, you
can pass the version number to apt-get (eg. `apt-get install
yarn=0.17.8-1`). If you use the install script, it can take a version as a
command line argument (eg. `./install.sh --version 0.17.8`).

Locking down the Yarn version used on CI servers is a good idea.

Sent from my phone.

On Nov 28, 2016 1:10 PM, "Pete Bacon Darwin" <notifications@github.com>
wrote:

> @Daniel15 <https://github.com/Daniel15> - thanks for the feedback about
> installing Yarn. Do you have any input on how we could lock down the
> version of Yarn that we are using to install our dependencies?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/15435#issuecomment-263394981>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFnHaJUypHpVKIYIjIGTinPlg9J47Duks5rC0NEgaJpZM4K8jcz>
> .
>

mgol commented on scripts/travis/before_build.sh:
```diff
@@ -2,6 +2,8 @@
 
 set -e
 
+yarn global add grunt-cli
```
Let's lock the version. ;-)

petebacondarwin replied:
locked!

petebacondarwin commented:
@Daniel15 - thanks. I think we are going to go with:

```
curl -o- -L https://raw.githubusercontent.com/yarnpkg/yarn/v0.17.9/scripts/install-latest.sh | bash -s -- --version 0.17.9
```

which ensures that the install script is also locked to the version being installed.

petebacondarwin committed:
docs(contribute): mention nvm-windows

Daniel15 commented:
@petebacondarwin - The version of `install-latest.sh` in the tag is missing some features, such as GPG verification and the `--version` flag. I don't think we cherry-picked any of the changes to the install script into the stable branch, as we expect everyone to use https://yarnpkg.com/install.sh rather than directly referencing the script on GitHub. Eventually the script will move out of the Yarn repo into the releases repo, where the Debian repo is also kept.


If you like, you could just have your own small script that downloads and extracts Yarn, based off the install-latest.sh script. Or, just grab https://yarnpkg.com/install.sh and check it into your repo.

mgol commented:
@Daniel15 Pete & me have tried using the install script but it doesn't work... I've tried downloading it & evaluating:

```shell
$ curl -o- -L https://yarnpkg.com/install.sh | bash -s -- --version 0.17.9
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    94  100    94    0     0    205      0 --:--:-- --:--:-- --:--:--   206
100  6074  100  6074    0     0   7515      0 --:--:-- --:--:-- --:--:-- 43078
Installing Yarn!
bash: line 164: yarn: command not found
```

as well as downloading first & then using:

```shell
$ curl -O -L https://yarnpkg.com/install.sh
$ chmod +x install.sh
$ ./install.sh --version 0.17.9
Installing Yarn!
./install.sh: line 164: yarn: command not found
```

Daniel15 commented:
@mgol - Looks like there's a bug in the script when `~/.yarn` exists but there's no Yarn version installed in it. I just filed an issue for it: https://github.com/yarnpkg/yarn/issues/2061

petebacondarwin committed:
chore(travis): install grunt-cli globally to simplify scripts

petebacondarwin committed:
chore(grunt): check node, yarn and grunt-cli versions

If global versions of node, yarn or grunt-cli don't match what we expect then blow up.

gkalpak commented on scripts/travis/before_build.sh:
```diff
@@ -2,7 +2,7 @@
 set -e
 
-yarn global add grunt-cli
+yarn global add grunt-cli@1.2.0
```
AFAICT, this will use the latest 1.x release. Is that intended or should we use the [`--exact` option][1]?

[1]: https://yarnpkg.com/en/docs/cli/add#toc-yarn-add-exact-e

mgol replied:
This is weird, npm in this case would install the provided version. Also, Yarn docs are contradicting each other in the very same page:

> `yarn add package-name@1.2.3` installs a specific version of a package from the registry.

vs:

> For example, `yarn add foo@1.2.3` would accept version `1.9.1`, but `yarn add foo@1.2.3 --exact` would only accept version `1.2.3`.

@Daniel15 could you clarify which one is right?

mgol replied:
No, it's CLI, we should lock everything as much as possible.

gkalpak replied:
Trying this locally, it seems that it will install the specific version (i.e. `--exact` makes no difference).
Probably a docs issue.

gkalpak commented on scripts/travis/build.sh:
```diff
@@ -14,9 +14,9 @@ elif [ "$JOB" == "unit" ]; then
-grunt tests:docs --browsers="$BROWSERS" --reporters=dots
+  grunt test:promises-aplus
+  grunt test:unit --browsers="$BROWSERS" --reporters=dots
+  grunt tests:docs --browsers="$BROWSERS" --reporters=dots
```
Good indentation! 😛

petebacondarwin committed:
chore(dependencies): USE YARN (and node 6)

gkalpak commented on Gruntfile.js:
```diff
@@ -10,26 +10,63 @@ var path = require('path');
+// Node.js version checks
+if (!semver.satisfies(process.version, pkg.engines.node)) {
+  reportOrFail('Invalid node version (' + process.version + '). ' +
+               'Please use a version that satisfies ' + pkg.engines.node);
```
Nit: Missing trailing period :grin:
Nit: I would make the message a little "softer". For example, the build will (hopefully) work fine with v7+; calling it an "invalid version" sounds "strong".

petebacondarwin replied:
You're too kind! If you feel strongly about it I will let you fix up the messaging in a future commit.

gkalpak commented on Gruntfile.js:
```diff
@@ -10,26 +10,63 @@ var path = require('path');
+var currentYarnVersion = exec('yarn --version', {silent: true}).stdout.trim();
+if (!semver.satisfies(currentYarnVersion, expectedYarnVersion)) {
+  reportOrFail('Invalid yarn version (' + currentYarnVersion + '). ' +
+               'Please use a version that satisfies ' + expectedYarnVersion);
```
Nit: Missing trailing period.

gkalpak commented on Gruntfile.js:
```diff
@@ -10,26 +10,63 @@ var path = require('path');
+} else {
+  if (!semver.satisfies(match[1], expectedGruntVersion)) {
+  reportOrFail('Invalid grunt-cli version (' + match[1] + '). ' +
+               'Please use a version that satisfies ' + expectedGruntVersion);
```
Nit: Missing trailing period.

gkalpak reviewed (approved):
LGTM
